src/runtime/hexagon/hexagon_device_api.cc

```diff
@@ -0,0 +1,158 @@
+    case kWarpSize:
+      return "kWarpSize";
+    case kMaxSharedMemoryPerBlock:
+      return "kMaxSharedMemoryPerBlock";
+    case kComputeVersion:
+      return "kComputeVersion";
+    case kDeviceName:
+      return "kDeviceName";
+    case kMaxClockRate:
+      return "kMaxClockRate";
+    case kMultiProcessorCount:
+      return "kMultiProcessorCount";
+    case kMaxThreadDimensions:
+      return "kMaxThreadDimensions";
+    default:
+      break;
+  }
+
+  std::stringstream ss;
+  ss << "<unknown attr, kind=" << kind << '>';
+  return ss.str();
+}
+
+// HexagonDeviceAPI.
+
+inline void HexagonDeviceAPI::SetDevice(TVMContext ctx) {}
+
+inline void HexagonDeviceAPI::GetAttr(TVMContext ctx, DeviceAttrKind kind,
+                                      TVMRetValue* rv) {
+  if (kind == kExist) *rv = 1;
+}
+
+inline void* HexagonDeviceAPI::AllocDataSpace(TVMContext ctx, size_t nbytes,
+                                              size_t alignment,
+                                              TVMType type_hint) {
+#ifdef __ANDROID__
+  __android_log_print(ANDROID_LOG_VERBOSE, "TVM",
+                      "Allocating memory for bytes = %zu for device:%d\n",
+                      nbytes, ctx.device_id);
+#endif
+  return hexagon::HW.get(ctx.device_id)->alloc(nbytes, alignment);
+}
+
+inline void HexagonDeviceAPI::FreeDataSpace(TVMContext ctx, void* ptr) {
+  hexagon::HW.get(ctx.device_id)->free(ptr);
+}
+
+inline void HexagonDeviceAPI::CopyDataFromTo(
+    const void* from, size_t from_offset, void* to, size_t to_offset,
+    size_t num_bytes, TVMContext ctx_from, TVMContext ctx_to,
+    TVMType type_hint, TVMStreamHandle stream) {
+  const char* src = static_cast<const char*>(from) + from_offset;
+  char* dst = static_cast<char*>(to) + to_offset;
+
+  auto is32bit = [](const void* p) {
+    return p == reinterpret_cast<const void*>(uint32_t(uintptr_t(p)));
+  };
+  (void)is32bit;
+  if (ctx_from.device_type == ctx_to.device_type) {
+    if (ctx_from.device_type == kDLCPU) {
+      memmove(dst, src, num_bytes);
+    } else if (static_cast<int>(ctx_from.device_type) == kDLHexagon) {
+      assert(ctx_from.device_id == ctx_to.device_id);
+      assert(is32bit(dst) && is32bit(src));
+      hexagon::HW.get(ctx_from.device_id)->copy(dst, src, num_bytes);
+    }
+  } else {
+    if (ctx_from.device_type == kDLCPU) {
+      assert(static_cast<int>(ctx_to.device_type) == kDLHexagon);
+      assert(is32bit(dst));
+      hexagon::HW.get(ctx_to.device_id)->copy_to(dst, src, num_bytes);
+    } else {
+      assert(static_cast<int>(ctx_from.device_type) == kDLHexagon);
+      assert(ctx_to.device_type == kDLCPU);
+      assert(is32bit(src));
+      hexagon::HW.get(ctx_from.device_id)->copy_from(dst, src, num_bytes);
+    }
+  }
+}
+
+inline void HexagonDeviceAPI::StreamSync(TVMContext ctx,
+                                         TVMStreamHandle stream) {}
+
+TVM_REGISTER_GLOBAL("device_api.hexagon")
+    .set_body([](TVMArgs args, TVMRetValue* rv) {
```

Comment:
Use `set_body_typed` if possible

Reply:
I've tried it---`set_body_typed` cannot infer all types when the argument is a lambda.